### ISSUE TITLE
user functions have no way to modify external state?

### ISSUE THREAD

**Issue by mkujawa** [fastbuild/fastbuild#742]
I was experimenting with the new functions, and I couldn't figure out how to get them to have any external side effects besides calling built-in functions.  

```
.MyStruct = 
[
  .Memb1 = 'One'
  .Memb2 = { 'An', 'Array' }
]

function TestFunc()
{
  Print(.MyStruct)
}

TestFunc()
```

`Unknown variable  '.MyStruct'`

Am I missing something?

**Comment by ffulin** [comment]
v1.00 contains the first iteration of functions, and is rather basic. They likely can't solve for all the use-cases we'd ultimately want. I expect them to be improved over coming versions.

However, the choice to make functions not have side effects and to not have access to global state is a conscious one.

**Inputs**

All inputs must be passed in explicitly. In your (obviously simplified example) you can pass in .MyStruct to the function:
```C++
function TestFunc(.Arg)
{
    Printf(.Arg)
}
TextFunc(.MyStruct)
```

In the future, support for optional arguments will be added, which will allow functions to be much more flexible.

**Outputs**

Currently, the only "Outputs" are ultimately invoking built in functions (directly or via other user functions). You could likely build an entire config in a useful way with this alone, but to complement existing bffs by replacing small snippets, their use is a little limitted (you have to fully encapsulate a subset of functionality within a function)

It's very likely that functions will eventually be able to return values which would further extend their usefulness/flexibility.

**Comment by mkujawa** [comment]
What I wanted to do was `^MyStruct + .localChanges` but that didn't work, and I made my way backwards to the most simple `Print`.

Return values would definitely help.  I think I might prefer the ability to take a variable "by reference,"  if I had to pick just one way to get state out, since that way I can mutate multiple structs or arrays more easily (and avoids copies.)  I personally don't care much about optional parameters at this point.  Despite the tone of the question, I'm actually happy for the added structure of functions not having access to their parent scope.  (Once they have a way to return state.)

Another feature that could be useful is function pointers.  Might be best to think what the syntax for "by reference" and "function pointers" would be at the same time.

Brain storm proposal:
```
.MyObj = [ .Ctor = &TestFunc1 .Dtor = &TestFunc2 .Work = &TestFunc3 ]
Using (.MyObj)
{
    .Ctor( "hello" )
    .results = [ ] 
    .is_ok = .Work( &.results )
    .Dtor( "world" )
}

Thank you for your reply.  I'll make issues for tracking all three things mentioned here.
